### PR TITLE
DOC: unpin numpydoc, fix validation script #39688

### LINKED ISSUE #39688: DOC: doc-string validation broken
https://github.com/pandas-dev/pandas/pull/39685/checks?check_run_id=1857708224

```
==================================== ERRORS ====================================
__________ ERROR collecting scripts/tests/test_validate_docstrings.py __________
ImportError while importing test module '/home/runner/work/pandas/pandas/scripts/tests/test_validate_docstrings.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/share/miniconda/envs/pandas-dev/lib/python3.8/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
scripts/tests/test_validate_docstrings.py:5: in <module>
    import validate_docstrings
scripts/validate_docstrings.py:50: in <module>
    from numpydoc.validate import validate, Docstring  # isort:skip
E   ImportError: cannot import name 'Docstring' from 'numpydoc.validate' (/usr/share/miniconda/envs/pandas-dev/lib/python3.8/site-packages/numpydoc/validate.py)
=========================== short test summary info ============================
ERROR scripts/tests/test_validate_docstrings.py
!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
=============================== 1 error in 0.54s ===============================
Error: Process completed with exit code 2.
````

cc @pandas-dev/pandas-core 
can we pin numpydoc?

### PR DESCRIPTION
**Pull request by fangchenli**
- [x] closes #39688

### CODE REVIEW

**Comment by fangchenli** [comment]
The breaking change was from https://github.com/numpy/numpydoc/pull/302. 

After modifying our script according to the new API. There is an AttributeError error:

```
pandas.Series.array <property object at 0x1199e99f0>
Traceback (most recent call last):
  File "scripts/validate_docstrings.py", line 438, in <module>
    main(
  File "scripts/validate_docstrings.py", line 384, in main
    return print_validate_all_results(
  File "scripts/validate_docstrings.py", line 328, in print_validate_all_results
    result = validate_all(prefix, ignore_deprecated)
  File "scripts/validate_docstrings.py", line 298, in validate_all
    doc_info = pandas_validate(func_name)
  File "scripts/validate_docstrings.py", line 241, in pandas_validate
    result["examples_errs"] = doc.examples_errors
  File "scripts/validate_docstrings.py", line 161, in examples_errors
    for test in finder.find(self.raw_doc, self.name, globs=context):
  File "/opt/homebrew/Caskroom/miniforge/base/envs/pandas-dev/lib/python3.8/site-packages/numpydoc/validate.py", line 135, in name
    return '.'.join([self.obj.__module__, self.obj.__name__])
AttributeError: 'property' object has no attribute '__module__'
```

**Comment by jbrockmendel** [comment]
@fangchenli is this actionable?

**Comment by jreback** [comment]
cc @MarcoGorelli @simonjayhawkins

File: scripts/validate_docstrings.py [hunk 1]
```diff
@@ -145,10 +145,17 @@ def examples_errors(self):
...
-        for test in finder.find(self.raw_doc, self.name, globs=context):
-            f = io.StringIO()
-            runner.run(test, out=f.write)
-            error_msgs += f.getvalue()
+        name = None
+        try:
+            name = self.name
+        except AttributeError:
+            if not isinstance(self.obj, property):
+                name = type(self.obj).__name__
```
Comment (MarcoGorelli) [inline scripts/validate_docstrings.py:153]: this isn't covered by any of the tests in `scripts/validate_docstrings.py` - when could it happen?

File: scripts/validate_docstrings.py [hunk 2]
```diff
@@ -145,10 +145,17 @@ def examples_errors(self):
...
+        if name is not None:
```
Comment (MarcoGorelli) [inline scripts/validate_docstrings.py:154]: likewise, none of the tests cover when `name` is `None` - when could that happen?

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]

**Comment by mroeschke** [comment]
This PR has appeared to have gone stale. Feel free to reopen if you're still interested in working on this.

**Pull request closed without merging**